### PR TITLE
Write basic documentation in README

### PR DESCRIPTION
**Pull request by jviotti**
Signed-off-by: Juan Cruz Viotti <jviotti@openmailbox.org>

### CODE REVIEW

**Comment by jviotti** [comment]
@alexandrosm Take a look at this. I wrote some basic documentation and explanation based on a fictitious example.

File: README.md [hunk 1]
```diff
@@ -12,11 +16,322 @@ Install `reconfix` by running:
...
+image. This file contains a set of key-value pairs following the [ini] file
+format. For example:
+
+```txt
+gpu_mem=192
+hdmi_drive=2
+hdmi_mode=82
+```
+
+Lets pretend we wrote a YouTube video player application for our Raspberry Pi,
```
Comment (WasabiFan) [inline README.md]: Contraction -- "Let's"

File: README.md [hunk 2]
```diff
@@ -12,11 +16,322 @@ Install `reconfix` by running:
...
+{
+  "user": "johndoe",
+  "apiKey": "xxxx-yyyy-zzzz",
+  "quality": 720,
+  "fullscreen": true,
+  "annotations": false
+}
+```
+
+Finally, lets say we've configured our Raspberry Pi to read network
```
Comment (WasabiFan) [inline README.md]: let's

File: README.md [hunk 3]
```diff
@@ -12,11 +16,322 @@ Install `reconfix` by running:
...
+Type = wifi
+Name = mynetwork
+Passphrase = secret
+Nameservers = 8.8.8.8,8.8.4.4
+```
+
+We want to build a tool that allows us to read configuration from all the files
+we care about, modify it, and write it back, but the problem is that the
+settings live in different files, which follow different formats. We can of
+course write custom code to handle this small example, but Reconfix allow us to
```
Comment (WasabiFan) [inline README.md]: "allows"

File: README.md [hunk 4]
```diff
@@ -12,11 +16,322 @@ Install `reconfix` by running:
...
+
+### Phases
+
+In order to provide bi-directional transformation of data, Reconfix makes use
+of different transformation phases.
+
+#### Dry JSON
+
+This phase consists in a high-level configuration state out of all the files we
+care about. Based on the example we outlined at the beginning of this section, we could design our dry JSON to look something like this:
```
Comment (WasabiFan) [inline README.md]: Hmmm... I think this line got a bit over-excited :laughing: The line wrapping took a vacation.

File: README.md [hunk 5]
```diff
@@ -12,11 +16,322 @@ Install `reconfix` by running:
...
+tackle the problem in a declarative way.
+
+### Phases
+
+In order to provide bi-directional transformation of data, Reconfix makes use
+of different transformation phases.
+
+#### Dry JSON
+
+This phase consists in a high-level configuration state out of all the files we
```
Comment (WasabiFan) [inline README.md]: _consists in_ should probably be _consists of_, and then there's a verb missing somewhere else

File: README.md [hunk 6]
```diff
@@ -12,11 +16,322 @@ Install `reconfix` by running:
...
+  }
+]
+```
+
+Notice how this schema defines a relationship between the structure of the dry
+JSON, and the structure of the wet JSON.
+
+### Wet JSON <-> Files JSON
+
+A schema to go from wet JSON to files JSON, and viceversa, would look something
```
Comment (WasabiFan) [inline README.md]: "vice versa" is a phrase with two words

**Review by WasabiFan — COMMENTED** [review]
Just a few minor typos and style/grammatical things I noticed while reading this. Super interesting!

**Comment by lurch** [comment]
IMHO it probably makes sense to separate the low-level partition-labelling / numbering part of this specification out of reconfix?
I think reconfix should only deal with reading and writing configuration from/to 'regular filesystem paths' (i.e. imagine that it's running on a live system (which I can easily imagine being done for realz), so `config.txt` should be specified as `/boot/config.txt`).

The partition-handling stuff (i.e. knowing that `/boot` maps to a FAT filesystem on partition 1, and all other paths map to an ext4 filesystem on partition 2) should be handled by some other process. E.g. maybe this "other process" could mount partition 2 at `/tmp/conftest` and mount partition 1 at `/tmp/conftest/boot` and then run something like `reconfix --prefix=/tmp/conftest`.
Or if you wanted to do it a single partition at a time (e.g. mounting each partition to `/tmp/conftest` and unmounting afterwards), maybe you could do something like `reconfix --prefix=/tmp/conftest --include=/boot --strip=/boot` for partition 1 and `reconfix --prefix=/tmp/conftest --include=/ --exclude=/boot` for partition 2 ?

File: README.md [hunk 7]
```diff
@@ -12,11 +16,322 @@ Install `reconfix` by running:
...
+
+Now that we understand the different transformation phases that Reconfix
+implements, we can talk about schemas. Schemas are user-defined declarative
+JSON files that help Reconfix to bi-directionally move between the phases
+described above.
+
+#### Dry JSON <-> Wet JSON
+
+This schema allows Reconfix to go from the dry JSON to the wet JSON, and
+viceversa. This is a valid schema for our YouTube player example:
```
Comment (alexandrosm) [inline README.md]: Vice versa is two words

**Comment by lurch** [comment]
Might also be worth clarifying if reconfix is always intended to (over)write the whole config file, or if it supports "merging" new config snippets into existing files?

File: README.md (hunk 1 above)
Comment (jviotti) [reply README.md]: Fixed!

File: README.md (hunk 2 above)
Comment (jviotti) [reply README.md]: Fixed!

File: README.md (hunk 3 above)
Comment (jviotti) [reply README.md]: Fixed!

File: README.md (hunk 5 above)
Comment (jviotti) [reply README.md]: Fixed!

File: README.md (hunk 4 above)
Comment (jviotti) [reply README.md]: Haha, I don't think I understand the comment entirely though. What's the fix?

File: README.md (hunk 6 above)
Comment (jviotti) [reply README.md]: Done!

File: README.md (hunk 7 above)
Comment (jviotti) [reply README.md]: Fixed!

**Comment by jviotti** [comment]
> IMHO it probably makes sense to separate the low-level partition-labelling / numbering part of this specification out of reconfix?

The problem is that the mount-point is not known in advance. If you're dealing with an `*.img`, you have no idea what partition will be mounted as `/boot` at runtime until you actually run the thing.

> Might also be worth clarifying if reconfix is always intended to (over)write the whole config file, or if it supports "merging" new config snippets into existing files?

Reconfix merges at the moment. I'll include that in the docs.

File: README.md (hunk 4 above)
Comment (WasabiFan) [reply README.md]: This line is a few hundred characters long, while others seem to be wrapped earlier. Could just be the GitHub viewer though.

Comment (jviotti) [reply README.md]: Yeah, looks like a GH issue:

screenshot 2016-12-14 12 56 00

**Comment by lurch** [comment]
> The problem is that the mount-point is not known in advance.

Yeah, that's why I also added the "Or if you wanted to do it a single partition at a time" section, so that you can tweak it to work with different mountpoints ;)

>  If you're dealing with an `*.img`, you have no idea what partition will be mounted as `/boot` at runtime until you actually run the thing.

In the specific case of dealing with an `*.img`, I think the expectation is that it's the person creating the `*.img` who is also the person creating the reconfix files?

I can see _why_ you're wanting to include partition information inside the reconfix files (given the direction that this project has come from), but my gut feeling is still that that level of information belongs in a different tool, and reconfix itself should only be expected to work at the filesystem level. Would optionally using filepaths like `/partition1/config.txt` and `/partition2/etc/hostname` alleviate both our concerns?

> Reconfix merges at the moment.

So how does that work for config files that can have "lists of sections" e.g. a list of wifi-credentials? How would reconfix know when to replace an existing entry with a new entry (e.g. if you're changing the password, or if you only want your list of wifi-credentials to allow storage of a single entry), and when to append the new entry?

**Comment by jviotti** [comment]
> In the specific case of dealing with an *.img, I think the expectation is that it's the person creating the *.img who is also the person creating the reconfix files?

I think that's a good assumption, and I'm very interested in dropping partitions for something more readable, like paths. The problem I see is that even if you're the image vendor and know that say the 1st partition will be mounted at `/boot`, you still need to be able to write to the image even when its not running (say to a `.img` or an SDCard at `/dev/sda` in your laptop), and it that context, the mount-point is completely useless -- e.g: get a ResinOS image and write to `/boot/foobar`. How do we fetch information from the image itself that `/boot` is partition 1?

> So how does that work for config files that can have "lists of sections" e.g. a list of wifi-credentials? How would reconfix know when to replace an existing entry with a new entry (e.g. if you're changing the password, or if you only want your list of wifi-credentials to allow storage of a single entry), and when to append the new entry?

Array merging is not something we dealt yet, and I'm sure that if you give it a go you'll find some weird bugs in there, so its definitely something to polish.

File: README.md [hunk 8]
```diff
@@ -12,11 +16,326 @@ Install `reconfix` by running:
...
+something like this:
+
+```json
+{
+  "boot_config_txt": {
+    "type": "ini",
+    "location": {
+      "path": "config.txt",
+      "partition": {
+        "primary: 1"
```
Comment (stefan-mihaila) [inline README.md]: minor typo: `"primary": 1`

File: README.md [hunk 9]
```diff
@@ -12,11 +16,326 @@ Install `reconfix` by running:
...
+      }
+    }
+  },
+  "etc_pitube_json": {
+    "type": "json",
+    "location": {
+      "path": "/etc/pitube.json",
+      "partition": {
+        "primary: 2"
```
Comment (stefan-mihaila) [inline README.md]: `"primary": 2`

File: README.md [hunk 10]
```diff
@@ -12,11 +16,326 @@ Install `reconfix` by running:
...
+      }
+    }
+  },
+  "etc_network": {
+    "type": "ini",
+    "location": {
+      "path": "/etc/network",
+      "partition": {
+        "primary: 2"
```
Comment (stefan-mihaila) [inline README.md]: `"primary": 2`

File: README.md (hunk 8 above)
Comment (jviotti) [reply README.md]: Great catches @stefan-mihaila ! I've fixed them all :)

File: README.md [hunk 11]
```diff
@@ -12,11 +16,326 @@ Install `reconfix` by running:
 $ npm install --save reconfix
 ```
 
+Documentation
+-------------
+
+Reconfix can be hard to grok. We'll discuss a concrete example in depth in
```
Comment (alexellis) [inline README.md:22]: > Reconfix can be hard to grok

Maybe that could come across condescending to some.. could reword?

**Comment by alexellis** [comment]
Huge improvement to see this documented.

**Comment by jviotti** [comment]
Closing as this has been superseded by the rust implementation.

**Pull request closed without merging**